### PR TITLE
refactor(voucher): remove billing details from voucher pmd

### PR DESCRIPTION
**Pull request by swangi-kumari**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement
- [x] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description
This PR includes changes to use  `payment_method_data.billing.first_name` ,`payment_method_data.billing.last_name` ,`payment_method_data.billing.phone.number` and `payment_method_data.billing.email` instead of `payment_method_data.[Voucher->AlfamartVoucherData].{first_name,last_name_email}` , `payment_method_data.[Voucher->IndomaretVoucherData].{first_name,last_name_email}` and `payment_method_data.[Voucher->JCSVoucherData].{first_name,last_name_email,phone_number}`at the connector level. The backwards compatibility is provided.

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?
1. Create a Boleto Payment Request via Adyen
```
{
    "amount": 2000,
    "currency": "BRL",
    "profile_id": "{{payment_profile_id}}",
    "confirm": true,
    "capture_method": "automatic",
    "capture_on": "2022-09-10T10:11:12Z",
    "amount_to_capture": 2000,
    "customer_id": "StripeCustomer",
    "email": "guest@example.com",
    "name": "John Doe",
    "phone": "999999999",
    "phone_country_code": "+1",
    "description": "Its my first payment request",
    "authentication_type": "no_three_ds",
    "return_url": "https://google.com",
    "payment_method": "voucher",
    "payment_method_type": "boleto",
    "payment_method_data": {
        "voucher": {
            "boleto": {
                "social_security_number": "56861752509"
            }
        }
    },
    "billing": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "BR",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "8056594427",
            "country_code": "+91"
        }
    },
    "shipping": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "US",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "8056594427",
            "country_code": "+91"
        }
    },
    "statement_descriptor_name": "joseph",
    "statement_descriptor_suffix": "JS",
    "metadata": {
        "udf1": "value1",
        "new_customer": "true",
        "login_date": "2019-09-10T10:11:12Z"
    }
}
```
Response
```
{
    "payment_id": "pay_hxmkR4zesMklFt6n8WwF",
    "merchant_id": "merchant_1713344003",
    "status": "requires_customer_action",
    "amount": 2000,
    "net_amount": 2000,
    "amount_capturable": 2000,
    "amount_received": null,
    "connector": "adyen",
    "client_secret": "pay_hxmkR4zesMklFt6n8WwF_secret_fGyotakujxfxB8vFu5CV",
    "created": "2024-04-17T09:01:22.583Z",
    "currency": "BRL",
    "customer_id": "StripeCustomer",
    "customer": {
        "id": "StripeCustomer",
        "name": "John Doe",
        "email": "guest@example.com",
        "phone": "999999999",
        "phone_country_code": "+1"
    },
    "description": "Its my first payment request",
    "refunds": null,
    "disputes": null,
    "mandate_id": null,
    "mandate_data": null,
    "setup_future_usage": null,
    "off_session": null,
    "capture_on": null,
    "capture_method": "automatic",
    "payment_method": "voucher",
    "payment_method_data": {
        "voucher": {},
        "billing": null
    },
    "payment_token": null,
    "shipping": {
        "address": {
            "city": "San Fransico",
            "country": "US",
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "zip": "94122",
            "state": "California",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "8056594427",
            "country_code": "+91"
        },
        "email": null
    },
    "billing": {
        "address": {
            "city": "San Fransico",
            "country": "BR",
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "zip": "94122",
            "state": "California",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "8056594427",
            "country_code": "+91"
        },
        "email": null
    },
    "order_details": null,
    "email": "guest@example.com",
    "name": "John Doe",
    "phone": "999999999",
    "return_url": "https://google.com/",
    "authentication_type": "no_three_ds",
    "statement_descriptor_name": "joseph",
    "statement_descriptor_suffix": "JS",
    "next_action": {
        "type": "display_voucher_information",
        "voucher_details": {
            "expires_at": 1713657600000,
            "reference": "03399.02447 50485.193010 27491.201011 1 96940000002000",
            "download_url": "https://test.adyen.com/hpp/generationBoleto.shtml?data=BQABAQAeOGdBuA26iNaifwOpm%2FF9orkr16oA1nXFyQ0d5tjl%2FW21mNMS9FBfwVedscbbyA1ecBTEIXFgZCXtBZGvfTC1JyP0Thv4D5mM7aYsmlRgOcUSrw%2FhwMIxMrbA%2BeY6B9jQoddl9E1EIGdq6iqYELpF8H0yH833o%2B4U9l8pu%2Fxa8wisamOJpnbpWwW3yyBMz8If7FQqxkM6vKLARsaOKkYx53Tw9%2BeAhrBGx91uNZnRuZ9eEZd0%2B5056xWGtvcPoMlElX4UmvLUOU94I7nL8alPs7s0oJN4r3IFxq5WflBTNVx5fgYV5a7ikKK9jQylYNxgpvuDevlpC4iRlmHM%2BIl0DHkYiEXm%2FlmQzKvIYgAApslW36JCe29%2FiiK861vjFxcgIhc5%2FA8wE7rtn2hFH8g7d9j7BC6vlITk9ZEOZJQSB7zrta2fpahoHyPgRJ7VsuX9szgc27mqE%2FZMLRD1Iu7LocYmenaqdZzTSliZz4mfRDWbHxLYdAt%2FuJOoWwi3wMmHXDDXXyi4au%2BxvnaVr44TaCgvx5dn0YOJ%2FSbnKnvSTmHVntCHwWG3W4QtYxAO0EPUwfrMNkRwWWmETyApZO2tjPDfVfbau%2BeO7omqUwLBk0dlu8F3fB7hdXTwjRjXTmuodlR8CQ%3D%3D",
            "instructions_url": null
        }
    },
    "cancellation_reason": null,
    "error_code": null,
    "error_message": null,
    "unified_code": null,
    "unified_message": null,
    "payment_experience": null,
    "payment_method_type": "boleto",
    "connector_label": null,
    "business_country": null,
    "business_label": "default",
    "business_sub_label": null,
    "allowed_payment_method_types": null,
    "ephemeral_key": {
        "customer_id": "StripeCustomer",
        "created_at": 1713344482,
        "expires": 1713348082,
        "secret": "epk_8673b49bd8284511a6af553c4b9969c7"
    },
    "manual_retry_allowed": null,
    "connector_transaction_id": "D94DSH65VBM6GVT5",
    "frm_message": null,
    "metadata": {
        "udf1": "value1",
        "login_date": "2019-09-10T10:11:12Z",
        "new_customer": "true"
    },
    "connector_metadata": null,
    "feature_metadata": null,
    "reference_id": "D94DSH65VBM6GVT5",
    "payment_link": null,
    "profile_id": "pro_WTCdm0asJJ6PWtPdQY3K",
    "surcharge_details": null,
    "attempt_count": 1,
    "merchant_decision": null,
    "merchant_connector_id": "mca_I5Jrrjdt8orLVTl5zklI",
    "incremental_authorization_allowed": null,
    "authorization_count": null,
    "incremental_authorizations": null,
    "external_authentication_details": null,
    "external_3ds_authentication_attempted": false,
    "expires_on": "2024-04-17T09:16:22.583Z",
    "fingerprint": null,
    "browser_info": null,
    "payment_method_id": null,
    "payment_method_status": null,
    "updated": "2024-04-17T09:01:24.734Z"
}
```
_______________________________________________
2. Create a Seven Eleven Payment Request via Adyen
```
{
    "amount": 2000,
    "currency": "JPY",
    "profile_id": "{{payment_profile_id}}",
    "confirm": true,
    "capture_method": "automatic",
    "capture_on": "2022-09-10T10:11:12Z",
    "amount_to_capture": 2000,
    "customer_id": "StripeCustomer",
    "email": "guest123@example1.com",
    "name": "John Doe",
    "phone": "999999999",
    "phone_country_code": "+1",
    "description": "Its my first payment request",
    "authentication_type": "no_three_ds",
    "return_url": "https://google.com",
    "payment_method": "voucher",
    "payment_method_type": "seven_eleven",
    "payment_method_data": {
        "voucher": {
            "seven_eleven": {
                "first_name": "Some",
                "last_name": "One",
                "email": "someone@something.com",
                "phone_number": "1234567890"
            }
        }
    },
    "billing": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "BR",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "email": "guest@example.com",
        "phone": {
            "number": "8056594427",
            "country_code": "+91"
        }
    },
    "shipping": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "US",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "8056594427",
            "country_code": "+91"
        }
    },
    "statement_descriptor_name": "joseph",
    "statement_descriptor_suffix": "JS",
    "metadata": {
        "udf1": "value1",
        "new_customer": "true",
        "login_date": "2019-09-10T10:11:12Z"
    }
}
```
Response
```
{
    "payment_id": "pay_ln5eGTMWp49vWgW24fwM",
    "merchant_id": "merchant_1713344003",
    "status": "requires_customer_action",
    "amount": 2000,
    "net_amount": 2000,
    "amount_capturable": 2000,
    "amount_received": null,
    "connector": "adyen",
    "client_secret": "pay_ln5eGTMWp49vWgW24fwM_secret_0zuUW5oXcnnnT6LNIvZS",
    "created": "2024-04-17T09:07:56.217Z",
    "currency": "JPY",
    "customer_id": "StripeCustomer",
    "customer": {
        "id": "StripeCustomer",
        "name": "John Doe",
        "email": "guest123@example1.com",
        "phone": "999999999",
        "phone_country_code": "+1"
    },
    "description": "Its my first payment request",
    "refunds": null,
    "disputes": null,
    "mandate_id": null,
    "mandate_data": null,
    "setup_future_usage": null,
    "off_session": null,
    "capture_on": null,
    "capture_method": "automatic",
    "payment_method": "voucher",
    "payment_method_data": {
        "voucher": {},
        "billing": null
    },
    "payment_token": null,
    "shipping": {
        "address": {
            "city": "San Fransico",
            "country": "US",
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "zip": "94122",
            "state": "California",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "8056594427",
            "country_code": "+91"
        },
        "email": null
    },
    "billing": {
        "address": {
            "city": "San Fransico",
            "country": "BR",
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "zip": "94122",
            "state": "California",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "8056594427",
            "country_code": "+91"
        },
        "email": "guest@example.com"
    },
    "order_details": null,
    "email": "guest123@example1.com",
    "name": "John Doe",
    "phone": "999999999",
    "return_url": "https://google.com/",
    "authentication_type": "no_three_ds",
    "statement_descriptor_name": "joseph",
    "statement_descriptor_suffix": "JS",
    "next_action": {
        "type": "redirect_to_url",
        "redirect_to_url": "http://localhost:8080/payments/redirect/pay_ln5eGTMWp49vWgW24fwM/merchant_1713344003/pay_ln5eGTMWp49vWgW24fwM_1"
    },
    "cancellation_reason": null,
    "error_code": null,
    "error_message": null,
    "unified_code": null,
    "unified_message": null,
    "payment_experience": null,
    "payment_method_type": "seven_eleven",
    "connector_label": null,
    "business_country": null,
    "business_label": "default",
    "business_sub_label": null,
    "allowed_payment_method_types": null,
    "ephemeral_key": {
        "customer_id": "StripeCustomer",
        "created_at": 1713344876,
        "expires": 1713348476,
        "secret": "epk_102bca474fad488ca5ecbff5d7c8ddec"
    },
    "manual_retry_allowed": null,
    "connector_transaction_id": "VJ7F358R73S6GVT5",
    "frm_message": null,
    "metadata": {
        "udf1": "value1",
        "login_date": "2019-09-10T10:11:12Z",
        "new_customer": "true"
    },
    "connector_metadata": null,
    "feature_metadata": null,
    "reference_id": "VJ7F358R73S6GVT5",
    "payment_link": null,
    "profile_id": "pro_WTCdm0asJJ6PWtPdQY3K",
    "surcharge_details": null,
    "attempt_count": 1,
    "merchant_decision": null,
    "merchant_connector_id": "mca_I5Jrrjdt8orLVTl5zklI",
    "incremental_authorization_allowed": null,
    "authorization_count": null,
    "incremental_authorizations": null,
    "external_authentication_details": null,
    "external_3ds_authentication_attempted": false,
    "expires_on": "2024-04-17T09:22:56.217Z",
    "fingerprint": null,
    "browser_info": null,
    "payment_method_id": null,
    "payment_method_status": null,
    "updated": "2024-04-17T09:08:01.370Z"
}
```

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/api_models/src/payments.rs [hunk 1]
```diff
@@ -1892,26 +1892,26 @@ impl GetAddressFromPaymentMethodData for BankRedirectData {
 pub struct AlfamartVoucherData {
     /// The billing first name for Alfamart
     #[schema(value_type = String, example = "Jane")]
-    pub first_name: Secret<String>,
+    pub first_name: Option<Secret<String>>,
```
Comment (Narayanbhat166) [inline crates/api_models/src/payments.rs:1895]: can you also change the schema type, this should be optional

File: crates/api_models/src/payments.rs [hunk 2]
```diff
@@ -1892,26 +1892,26 @@ impl GetAddressFromPaymentMethodData for BankRedirectData {
...
     /// The billing second name for Alfamart
     #[schema(value_type = String, example = "Doe")]
     pub last_name: Option<Secret<String>>,
     /// The Email ID for Alfamart
     #[schema(value_type = String, example = "example@me.com")]
-    pub email: Email,
+    pub email: Option<Email>,
```
Comment (Narayanbhat166) [inline crates/api_models/src/payments.rs:1901]: same here, please change the `value_type`

File: crates/api_models/src/payments.rs [hunk 3]
```diff
@@ -2461,21 +2461,21 @@ impl GetAddressFromPaymentMethodData for VoucherData {
...
                     ..AddressDetails::default()
                 }),
                 phone: None,
-                email: Some(voucher_data.email.clone()),
+                email: voucher_data.email.clone(),
             }),
             Self::Indomaret(voucher_data) => Some(Address {
                 address: Some(AddressDetails {
-                    first_name: Some(voucher_data.first_name.clone()),
+                    first_name: Some(voucher_data.first_name.clone()?),
```
Comment (Narayanbhat166) [inline crates/api_models/src/payments.rs:2473]: This question mark can be removed

File: crates/router/src/connector/adyen/transformers.rs [hunk 4]
```diff
@@ -734,14 +734,30 @@ pub enum OnlineBankingCzechRepublicBanks {
...
+            first_name: item.get_optional_billing_first_name().ok_or(
+                errors::ConnectorError::MissingRequiredField {
+                    field_name: "billing.first_name",
+                },
+            )?,
+            last_name: Some(item.get_optional_billing_last_name().ok_or(
+                errors::ConnectorError::MissingRequiredField {
+                    field_name: "billing.email",
+                },
+            )?),
```
Comment (Narayanbhat166) [inline crates/router/src/connector/adyen/transformers.rs:750]: is last name mandatory here?

File: crates/router/src/connector/adyen/transformers.rs [hunk 5]
```diff
@@ -734,14 +734,26 @@ pub enum OnlineBankingCzechRepublicBanks {
...
+            )?,
+            last_name: item.get_optional_billing_last_name(),
+            shopper_email: item.get_optional_billing_email().ok_or(
+                errors::ConnectorError::MissingRequiredField {
+                    field_name: "payment_method_data.billing.email",
+                },
+            )?,
+            telephone_number: item.get_optional_billing_phone_number().ok_or(
+                errors::ConnectorError::MissingRequiredField {
+                    field_name: "payment_method_data.billing.email",
```
Comment (Narayanbhat166) [inline crates/router/src/connector/adyen/transformers.rs:754]:
```suggestion
                    field_name: "payment_method_data.billing.phone.number",
```

**Review by Narayanbhat166 — DISMISSED** [review]

**Review by srujanchikke — DISMISSED** [review]

**Review by Narayanbhat166 — APPROVED** [review]

**Review by srujanchikke — APPROVED** [review]

**Pull request merged**